Userland/Libraries/LibWeb/Communication/WebSocket.cpp

```diff
@@ -0,0 +1,275 @@
+void WebSocket::on_close(u16 code, String reason, bool was_clean)
+{
+    // 1. Change the readyState attribute's value to CLOSED. This is handled by the Protocol's WebSocket
+    // 2. If [needed], fire an event named error at the WebSocket object. This is handled by the Protocol's WebSocket
+    dispatch_event(CloseEvent::create(EventNames::close, was_clean, code, reason));
+}
+
+// https://html.spec.whatwg.org/multipage/web-sockets.html#feedback-from-the-protocol
+void WebSocket::on_message(ByteBuffer message, bool is_text)
+{
+    if (m_websocket->ready_state() != Protocol::WebSocket::ReadyState::Open)
+        return;
+    if (is_text) {
+        auto text_message = String(ReadonlyBytes(message));
+        dispatch_event(MessageEvent::create(EventNames::message, text_message, url()));
+        return;
+    }
+    // type indicates that the data is Binary and binaryType is "blob"
+    // type indicates that the data is Binary and binaryType is "arraybuffer"
+    TODO();
+}
+
+bool WebSocket::dispatch_event(NonnullRefPtr<DOM::Event> event)
+{
+    return DOM::EventDispatcher::dispatch(*this, move(event));
+}
+
+JS::Object* WebSocket::create_wrapper(JS::GlobalObject& global_object)
+{
+    return wrap(global_object, *this);
+}
+
+#undef __ENUMERATE
+#define __ENUMERATE(attribute_name, event_name)                    \
+    void WebSocket::set_##attribute_name(HTML::EventHandler value) \
+    {                                                              \
+        set_event_handler_attribute(event_name, move(value));      \
+    }                                                              \
+    HTML::EventHandler WebSocket::attribute_name()                 \
+    {                                                              \
+        return get_event_handler_attribute(event_name);            \
+    }
+ENUMERATE_WEBSOCKET_EVENT_HANDLERS(__ENUMERATE)
+#undef __ENUMERATE
+
+void WebSocket::set_event_handler_attribute(const FlyString& name, HTML::EventHandler value)
+{
+    RefPtr<DOM::EventListener> listener;
+    if (!value.callback.is_null()) {
+        listener = adopt_ref(*new DOM::EventListener(move(value.callback)));
+    } else {
+        StringBuilder builder;
+        builder.appendff("function {}(event) {{\n{}\n}}", name, value.string);
+        auto parser = JS::Parser(JS::Lexer(builder.string_view()));
+        auto program = parser.parse_function_node<JS::FunctionExpression>();
+        if (parser.has_errors()) {
+            dbgln("Failed to parse script in event handler attribute '{}'", name);
+            return;
+        }
+        auto* function = JS::ScriptFunction::create(script_execution_context()->interpreter().global_object(), name, program->body(), program->parameters(), program->function_length(), nullptr, false, false);
+        VERIFY(function);
+        listener = adopt_ref(*new DOM::EventListener(JS::make_handle(static_cast<JS::Function*>(function))));
+    }
+    if (listener) {
```

Comment:
hmm, under what circumstances do we reach this with the RefPrt still being null?

Reply:
This method is actually a copy-paste from `GlobalEventHandlers.cpp` (I should mark it as such). I assume that it's not null biven the current code, but could become so if needed?

Reply:
oh ok, feel free to leave it as-is then.